File: swimming-contest-software/Server/src/main/java/network/AbstractServer.java
package network;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

public abstract class AbstractServer {
    private final int port;
    private ServerSocket server=null;
    protected static final Logger logger = LogManager.getLogger();

    public AbstractServer(int port) {
        this.port = port;
    }

    public void start() {
        try {
            server = new ServerSocket(port);
            while (true) {
                logger.info("Waiting for clients...");
                Socket client = server.accept();
                logger.info("Client connected...");
                processRequest(client);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            stop();
        }
    }

    public abstract void processRequest(Socket client);

    public void stop() {
        try {
            server.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Server/Network/AbstractServer.cs
using System.Net;
using System.Net.Sockets;
using log4net;

namespace Server.Network;

public abstract class AbstractServer
{
    protected static readonly ILog Logger = LogManager.GetLogger("Server");
    private readonly string _host;
    private readonly int _port;
    private TcpListener _socket;

    public AbstractServer(string host, int port)
    {
        _host = host;
        _port = port;
    }

    public void Start()
    {
        var ipAddress = IPAddress.Parse(_host);
        var ipEndPoint = new IPEndPoint(ipAddress, _port);
        _socket = new TcpListener(ipEndPoint);
        _socket.Start();
        Logger.Info("Server started successfully!");
        while (true)
        {
            Logger.Info("Waiting for clients...");
            var client = _socket.AcceptTcpClient();
            Logger.Info("Client connected!");
            ProcessRequest(client);
        }
    }

    public abstract void ProcessRequest(TcpClient client);
}